### ISSUE TITLE
[require-atomic-updates] False positive on reused loop variable name

### ISSUE THREAD

**Issue by fluggo** [typescript-eslint/typescript-eslint#3180]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/require-atomic-updates": "error"
  }
}
```

```TS
interface Entry {
  id: number;
}

async function foo(e: Entry | null) {
}

async function run() {
  const input: Entry[] = [];
  const props: Entry[] = [];

  for(const entry of input) {
    const prop = props.find(a => a.id === entry.id) || null;
    await foo(entry);
  }

  // This is flagged, but shouldn't be
  for(const entry of input) {
    const prop = props.find(a => a.id === entry.id) || null;
  }

  // This is not flagged
  for(const entry2 of input) {
    const prop = props.find(a => a.id === entry2.id) || null;
  }
}
```

tsconfig.json
```JSON
{
  "compileOnSave": false,
  "compilerOptions": {
    "baseUrl": "./",
    "declaration": false,
    "moduleResolution": "node",
    "alwaysStrict": true,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "strictFunctionTypes": true,
    "strictNullChecks": true,
    "sourceMap": true,
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": ["es2020"],
    "target":"es2020",
    "module": "commonjs",
    "esModuleInterop": true,
    "skipDefaultLibCheck": true,
  }
}
```

**Expected Result**

I expected no errors because each loop is using a different loop variable.

**Actual Result**

The second loop is flagged because it has the same variable name `entry` as the previous loop, but the variable used in the second loop is separate from the first.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.17.0` |
| `@typescript-eslint/parser`        | `4.17.0` |
| `TypeScript`                       | `4.2.3` |
| `ESLint`                           | `7.21.0` |
| `node`                             | `14.15.4` |

**Comment by fluggo** [comment]
So sorry, just realized I put this in the wrong project! Please close with extreme prejudice!
